### PR TITLE
Replace the header_links plugin with client-side generated anchors.

### PR DESCRIPTION
**Pull request by bryanbraun**
Fixes facebook/react#4124

### CODE REVIEW

File: docs/_layouts/default.html [hunk 1]
```diff
@@ -90,6 +90,32 @@
...
+        } else {
+          var textMethod = document.body.textContent ? "textContent" : "innerText";
+          var roughText = elements[i][textMethod];
+
+          // Regex rule for making the title URL-friendly.
+          tidyText = roughText.toLowerCase()
+                              .replace(/\s+/g, '-')
+                              .replace(/[^A-Za-z0-9\-_.]/g, '');
+
+          elements[i].setAttribute('id', tidyText);
```
Comment (Daniel15) [inline docs/_layouts/default.html]: Do the anchors actually work, since you're setting the ID after it's rendered? Does the browser correctly scroll down to the anchor when you open a fresh tab and go to a URL with an anchor?

File: docs/_layouts/default.html [hunk 2]
```diff
@@ -90,6 +90,32 @@
...
+          elementID = tidyText;
+        }
+        var anchor = '<a class="anchor" href="#' + elementID + '">#</a>';
+        elements[i].innerHTML = elements[i].innerHTML + anchor;
```
Comment (Daniel15) [inline docs/_layouts/default.html]: This is really inefficient as it needs to serialize the HTML (the read of `innerHTML`), then append text, then deserialize the text back into DOM nodes. I'd suggest creating a DOM node directly instead, something like this (untested):
```
var anchor = document.createElement('a');
anchor.className = 'anchor';
anchor.href = '#' + elementID;
elements[i].appendChild(anchor);
```

Comment (bryanbraun) [inline docs/_layouts/default.html]: Good thinking. I'll update this shortly.

File: docs/_layouts/default.html (hunk 1 above)
Comment (bryanbraun) [inline docs/_layouts/default.html]: The anchors do work. I did some testing across the various desktop browsers where I added anchors at various points in the loading process. I found that they all hop to IDs added after the DOM is fully loaded, though they differ in how much later in the process they will successfully jump to the anchor. In case you're curious, here's where each of the browsers fared on the test (the browser name is listed where they last successfully jumped to the anchor).

* **On DOM Ready (DOMContentLoaded)**
  * Safari
  * Chrome
  * Opera
* **On Page Fully Loaded (window.onload)**
  * Firefox
  * IE
* **After Arbitrary setTimeout**
  * - none -

In short, we're safe in adding the IDs where we have, at least for all the browsers I tested.

**Comment by bryanbraun** [comment]
OK, this is ready for review again. I also included a CSS adjustment that ensures the Sticky header doesn't obscure the heading being anchored to.

**Comment by lacker** [comment]
OK so this is fairly old, we redid a lot of the website since then, and there are merge conflicts. @bryanbraun does this PR still make sense and are you still interested in working on it? Or should we just deem it "timed out" and close it.

**Comment by gaearon** [comment]
Btw this is the oldest open PR in React repo :-)

**Comment by bryanbraun** [comment]
@lacker, thanks for following up. I'd like to try rebasing it later this week, and seeing if this kind of change still makes sense with the updated site. I'll let you know what I find by end of day Friday.

**Comment by bryanbraun** [comment]
@lacker, I've rebased the PR so it should apply cleanly now. The code is in the same place as it was in the original PR (before the closing body tag) but it can be moved if needed.

File: docs/css/react.scss [hunk 3]
```diff
@@ -122,6 +122,7 @@ h1, h2, h3, h4, h5, h6 {
 .hash-link {
   color: $mediumTextColor;
   display: none;
+  padding-left: 8px;
```
Comment (lacker) [inline docs/css/react.scss]: Is this a UI change or is this just to compensate for something that's happening in that Redcarpet library? If it is a UI change, would you mind posting a before and after screen shot to show what you're changing here?

File: docs/_layouts/default.html [hunk 4]
```diff
@@ -62,6 +62,36 @@
 
   <div id="fb-root"></div>
   <script>
+    (function() {
```
Comment (lacker) [inline docs/_layouts/default.html]: It seems inconsistent to inline this when things like `live_editor.js` are included as separate scripts, what do you think about refactoring this out similar to the other stuff?

I suspect people in the future will be digging through old code and trying to figure out what the heck this bit is doing, so would you mind adding just enough overall comments to explain what it is that this section of code is trying to accomplish?

File: docs/css/react.scss (hunk 3 above)
Comment (bryanbraun) [inline docs/css/react.scss]: It's compensating... changing how we add space between the title and hash-link. The UI should look the same:

hello_world_-_react_and_hello_world_-_react

File: docs/_layouts/default.html (hunk 4 above)
Comment (bryanbraun) [inline docs/_layouts/default.html]: Good call. I've made these updates in 1757114.

**Comment by bryanbraun** [comment]
Ok, that should be all the changes you requested. 👍

**Comment by lacker** [comment]
Thank you! This seems a bit less technical debty, in addition to working better with RSS, since I have a sneaking suspicion that React folks would prefer to write javascript than ruby.

**Comment by gaearon** [comment]
😮
